packages/foundation-models/src/azure-openai/azure-openai-stream.ts

```diff
@@ -0,0 +1,203 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { LineDecoder } from './azure-openai-line-decoder.js';
+import { SSEDecoder } from './azure-openai-sse-decoder.js';
+import type { ServerSentEvent } from './azure-openai-sse-decoder.js';
+import type { HttpResponse } from '@sap-cloud-sdk/http-client';
+
+const logger = createLogger({
+  package: 'foundation-models',
+  messageContext: 'azure-openai-stream'
+});
+
+type Bytes = string | ArrayBuffer | Uint8Array | Buffer | null | undefined;
+
+/**
+ * Stream implemented as an async iterable.
+ * @internal
+ */
+export class Stream<Item> implements AsyncIterable<Item> {
+  protected static fromSSEResponse<Item>(response: HttpResponse): Stream<Item> {
+    let consumed = false;
+
+    async function* iterator(): AsyncIterator<Item, any, undefined> {
+      if (consumed) {
+        throw new Error(
+          'Cannot iterate over a consumed stream, use `.tee()` to split the stream.'
+        );
+      }
+      consumed = true;
+      let done = false;
+      for await (const sse of _iterSSEMessages(response)) {
+        if (done) {
+          continue;
+        }
+
+        if (sse.data.startsWith('[DONE]')) {
+          done = true;
+          continue;
+        }
+
+        if (sse.event === null) {
+          let data;
+
+          try {
+            data = JSON.parse(sse.data);
+          } catch (e: any) {
+            logger.error(`Could not parse message into JSON: ${sse.data}`);
+            logger.error(`From chunk: ${sse.raw}`);
+            throw e;
+          }
+
+          if (data && data.error) {
+            throw new Error(data.error);
+          }
+
+          yield data;
+        } else {
+          let data;
+          try {
+            data = JSON.parse(sse.data);
+          } catch (e: any) {
+            logger.error(`Could not parse message into JSON: ${sse.data}`);
+            logger.error(`From chunk: ${sse.raw}`);
+            throw e;
+          }
+          // TODO: Is this where the error should be thrown?
+          if (sse.event === 'error') {
+            // throw new Error(data.error, data.message);
+            throw new Error(data.error);
+          }
+          yield { event: sse.event, data } as any;
+        }
+      }
+      done = true;
+    }
+
+    return new Stream(iterator);
+  }
+
+  constructor(public iterator: () => AsyncIterator<Item>) {}
+
+  [Symbol.asyncIterator](): AsyncIterator<Item> {
+    return this.iterator();
+  }
+}
+
+/**
+ * @internal
+ */
+export async function* _iterSSEMessages(
+  response: HttpResponse
+): AsyncGenerator<ServerSentEvent, void, unknown> {
+  if (!response.data) {
```

Comment:
I am actually in favour of this style, putting it within the function as this can be seen as a precondition / contract of this function. Otherwise, moving it into `iterator()` makes that function even longer.